datasets/mdd/mdd.py

```diff
@@ -0,0 +1,569 @@
+            # This defines the different columns of the dataset and their types
+            features=features,  # Here we define them above because they are different between the two configurations
+            # If there's a common (input, target) tuple from the features,
+            # specify them here. They'll be used if as_supervised=True in
+            # builder.as_dataset.
+            supervised_keys=None,
+            # Homepage of the dataset for documentation
+            homepage=_HOMEPAGE,
+            # License for the dataset if available
+            license=_LICENSE,
+            # Citation for the dataset
+            citation=_CITATION,
+        )
+
+    def _split_generators(self, dl_manager):
+        """Returns SplitGenerators."""
+        if self.config.name != "task4_reddit":
+            my_urls = ZIP_URL  # Cannot download just one single type as it is a compressed file.
+        else:
+            my_urls = REDDIT_URL
+        data_dir = dl_manager.download_and_extract(my_urls)
+        splits = [
+            datasets.SplitGenerator(
+                name=datasets.Split.TRAIN,
+                # These kwargs will be passed to _generate_examples
+                gen_kwargs={
+                    "filepath": os.path.join(data_dir, paths[self.config.name]["train"]),
+                },
+            ),
+            datasets.SplitGenerator(
+                name=datasets.Split.TEST,
+                # These kwargs will be passed to _generate_examples
+                gen_kwargs={
+                    "filepath": os.path.join(data_dir, paths[self.config.name]["test"]),
+                },
+            ),
+            datasets.SplitGenerator(
+                name=datasets.Split.VALIDATION,
+                # These kwargs will be passed to _generate_examples
+                gen_kwargs={
+                    "filepath": os.path.join(data_dir, paths[self.config.name]["dev"]),
+                },
+            ),
+        ]
+        if self.config.name == "task4_reddit":
+            splits += [
+                datasets.SplitGenerator(
+                    name=datasets.Split("cand_valid"),
+                    # These kwargs will be passed to _generate_examples
+                    gen_kwargs={
+                        "filepath": os.path.join(data_dir, paths[self.config.name]["cand_valid"]),
+                    },
+                ),
+                datasets.SplitGenerator(
+                    name=datasets.Split("cand_test"),
+                    # These kwargs will be passed to _generate_examples
+                    gen_kwargs={
+                        "filepath": os.path.join(data_dir, paths[self.config.name]["cand_test"]),
+                    },
+                ),
+            ]
+        return splits
+
+    def _generate_examples(self, filepath):
+        if "cand" not in filepath:
+            with open(filepath, encoding="utf-8") as f:
+                data = f.readlines()
```

Comment:
can you iterate over the lines instead of reading the full file to save memory ?